akka-pki/src/main/scala/akka/pki/pem/DERPrivateKeyLoader.scala

```diff
@@ -0,0 +1,130 @@
+import java.security.spec.PKCS8EncodedKeySpec
+import java.security.spec.RSAMultiPrimePrivateCrtKeySpec
+import java.security.spec.RSAOtherPrimeInfo
+import java.security.spec.RSAPrivateCrtKeySpec
+
+import akka.annotation.ApiMayChange
+import akka.pki.pem.PEMDecoder.DERData
+import com.hierynomus.asn1.ASN1InputStream
+import com.hierynomus.asn1.encodingrules.der.DERDecoder
+import com.hierynomus.asn1.types.constructed.ASN1Sequence
+import com.hierynomus.asn1.types.primitive.ASN1Integer
+
+final class PEMLoadingException(message: String, cause: Throwable) extends RuntimeException(message, cause) {
+  def this(msg: String) = this(msg, null)
+}
+
+object DERPrivateKeyLoader {
+
+  /**
+   * Converts the DER payload in [[PEMDecoder.DERData]] into a [[java.security.PrivateKey]]. The received DER
+   * data must be a valid PKCS#1 (identified in PEM as "RSA PRIVATE KEY") or non-ecnrypted PKCS#8 (identified
+   * in PEM as "PRIVATE KEY").
+   * @throws PEMLoadingException when the `derData` is for an unsupported format
+   */
+  @ApiMayChange
+  @throws[PEMLoadingException]("when the `derData` is for an unsupported format")
+  def load(derData: DERData): PrivateKey = {
+    derData.label match {
+      case "RSA PRIVATE KEY" =>
+        loadPkcs1PrivateKey(derData.bytes)
+      case "PRIVATE KEY" =>
+        loadPkcs8PrivateKey(derData.bytes)
+      case unknown =>
+        throw new PEMLoadingException(s"Don't know how to read a private key from PEM data with label [$unknown]")
+    }
+  }
+
+  private def loadPkcs1PrivateKey(bytes: Array[Byte]) = {
+    val derInputStream = new ASN1InputStream(new DERDecoder, bytes)
+    // Here's the specification: https://tools.ietf.org/html/rfc3447#appendix-A.1.2
+    val sequence = {
+      try {
+        derInputStream.readObject[ASN1Sequence]()
+      } finally {
+        derInputStream.close()
+      }
+    }
+    val version = getInteger(sequence, 0, "version").intValueExact()
+    if (version < 0 || version > 1) {
+      throw new IllegalArgumentException(s"Unsupported PKCS1 version: $version")
+    }
+    val modulus = getInteger(sequence, 1, "modulus")
+    val publicExponent = getInteger(sequence, 2, "publicExponent")
+    val privateExponent = getInteger(sequence, 3, "privateExponent")
+    val prime1 = getInteger(sequence, 4, "prime1")
+    val prime2 = getInteger(sequence, 5, "prime2")
+    val exponent1 = getInteger(sequence, 6, "exponent1")
+    val exponent2 = getInteger(sequence, 7, "exponent2")
+    val coefficient = getInteger(sequence, 8, "coefficient")
+
+    val keySpec = if (version == 0) {
+      new RSAPrivateCrtKeySpec(
+        modulus,
+        publicExponent,
+        privateExponent,
+        prime1,
+        prime2,
+        exponent1,
+        exponent2,
+        coefficient)
+    } else {
+      // Does anyone even use multi-primes? Who knows, maybe this code will never be used. Anyway, I guess it will work,
+      // the spec isn't exactly complicated.
```

Comment:
For posterity, I tried to write tests for this (there's a comment below), but as far as I can see, while the JDK does have the `RSAMultiPrimePrivateCrtKeySpec` to represent multiprime private keys, as far as I could see, the built in JSSE provider ignores the additional primes, so there's no way to test that using this parsed key spec works, and the built in PKCS8 parser throws an exception if you pass it a multi prime key, so I couldn't compare the output of this to a PKCS8 parsed key. The tests just test that multi prime keys are parsed without throwing an exception, whether what's parsed is correct is untested because there was no real way to test it. I suppose if you use a different JSSE provider, you might get multi prime support, and then this code will be useful if you use multiprime keys, but the fact that the JDK doesn't implement it suggests to me that it's a feature that's never used, so this code is possibly effectively dead code.